### ISSUE TITLE
NetNewsWire doesn't find some YouTube feeds

### ISSUE THREAD

**Issue by superlex** [brentsimmons/NetNewsWire#902]
Hello,
first of all thank you and congratulations for your amazing work :)

Then,
currently, it looks like NetNewsWire doesn't find YouTube feeds if the link contains channel id instead of user name.

**Example 1:**
Try to add
`https://www.youtube.com/user/CNN`
Result: ok

**Example 2:**
Try to add
`https://www.youtube.com/channel/UChjRM_qQAaOAiLNbOGbYcRA/`
Result: Feed not found

Warkaround: convert the link into
`https://www.youtube.com/feeds/videos.xml?channel_id=UChjRM_qQAaOAiLNbOGbYcRA`
Result: ok

**My environment:**
- NetNewsWire Version 5.0 (2609)
- macOS 10.14.6

**Comment by brentsimmons** [comment]
We’ll fix this for 5.0.1 if it’s a defect in the feed finder. If it’s not — if it would take a special case — we’ll punt till after 5.0.1.

**Comment by brentsimmons** [comment]
This is not, strictly speaking, a defect in the feed finder. That page doesn’t link to an RSS feed (or Atom or anything else).

Finding this feed would require doing a special case — doing some pattern matching on the URL and then constructing the likely feed URL.

I’m making this a feature request and removing it from the 5.0.1 milestone.

**Comment by superlex** [comment]
Ah ok, thank you!

**Comment by eivindml-solvr** [comment]
Would love this as well. As a workaround I created an AppleScript that creates the feed url, and opens it in NNW.

```javascript
var base = "https://www.youtube.com/feeds/videos.xml?channel_id="
var safari = Application('Safari')

var url = safari.windows[0].currentTab.url()
var res = url.split("/")
var id = res[res.length - 1]

safari.windows[0].currentTab.url = base + id
```

<img width="365" alt="Skjermbilde 2020-02-06 kl  12 56 33" src="https://user-images.githubusercontent.com/55379708/73935127-26db4580-48e0-11ea-9ab8-7c05e1fbad27.png">

**Comment by xxyzz** [comment]
Maybe we can port some code from NewPipeExtractor(https://github.com/TeamNewPipe/NewPipeExtractor/tree/dev/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube)

**Comment by fannar** [comment]
Youtube channels do have a RSS link, but NNW doesn't recognize it.

`<link rel="alternate" type="application/rss+xml" title="RSS" href="https://www.youtube.com/feeds/videos.xml?channel_id=UChjRM_qQAaOAiLNbOGbYcRA">`

**Comment by sayomelu** [comment]
> Youtube channels do have a RSS link, but NNW doesn't recognize it.
> 
> `<link rel="alternate" type="application/rss+xml" title="RSS" href="https://www.youtube.com/feeds/videos.xml?channel_id=UChjRM_qQAaOAiLNbOGbYcRA">`

@brentsimmons you can let nnw find youtube rss link via this html element.

**Comment by brentsimmons** [comment]
Sounds like it could be a bug in our code that looks for the `<link…` tag with the RSS URL.

**Comment by vincode-io** [comment]
@fannar Could you please supply us with the URL you found the `link` element in? I'm not seeing it when looking at the source for `https://www.youtube.com/user/TheVolvorocks` where I would expect to see it.

**Comment by xxyzz** [comment]
> I'm not seeing it when looking at the source for `https://www.youtube.com/user/TheVolvorocks` where I would expect to see it.

It's inside the `body` tag:
<img width="1021" alt="Screen Shot 2021-05-14 at 10 56 12" src="https://user-images.githubusercontent.com/21101839/118214788-031f8b80-b4a3-11eb-9874-d5bf3d2a69c9.png">

**Comment by brentsimmons** [comment]
I just did a workaround so that we’ll detect feed links in YouTube pages like https://www.youtube.com/user/TheVolvorocks

Our code stopped looking once it hit the body tag, since feed links should be in the head section. I created a special case for YouTube where it will continue scanning for feed links, instead of stopping, once it hits the body tag.

**Comment by ednl** [comment]
Have YouTube since gone one step further in removing all mention of RSS? ;( I'm not seeing any rss links on channel pages. I have a few old YouTube subscriptions in NNW so I can still add them manually by editing an exported list but that's not really convenient. E.g. I don't see any rss link in the source for https://www.youtube.com/channel/UCCxKPNMqjnqbxVEt1tyDUsA and consequently the toolbar extension doesn't find anything. But my old link works in NNW!

**Comment by honzajavorek** [comment]
This has got a bit worse by the fact YouTube has introduced handles for channels. Adding e.g. https://www.youtube.com/@DreamPrague now involves quite some manual labor to find out the ID as it's not in the URL anymore. It would be awesome if NetNewsWire supported magically constructing the RSS links out of the box if given the channel URL.

**Comment by Krinkle** [comment]
TLDR: This seems to be resolved and the issue can be closed?

YouTube outputs some invisible schema.org metadata inside a `<span>…</span>` within the HTML `<head>`. A valid HTML5 parser must treat a `<span>` before `<body>` as the _implicit_ end of the `<head>` element, and the _implicit_ start of the `<body>` element. 

I suspect this may've been the case two years ago in 2023 as well, and so YouTube perhaps did not actually output the `<link>` tag in the body in terms of the HTML text. But, it ends up there regardless when parsed into a valid DOM. This is a distinction without a difference as far as NetNewsWire is concerned, but it shows there is at least intent for it to be part of the `<head>`. In other words, YouTube knows and wants to put the link tag in the head, and don't have any product reasons or technical reasons to delay it to "late" in the body. As far as they're concerned, it's in the head, but they have a separate bug in the form of a `<span>` tag in the head breaking everything after it. You'd think Google, having launched schema.org, knows how to produce Schema.org metadata and how it needs to either be declared as JSON-LD in the head, or as HTML in the the body.

Example from today <https://www.youtube.com/channel/UChjRM_qQAaOAiLNbOGbYcRA>:

```html
<!DOCTYPE html>
<html …>
  <head …>
    <script …>…</script>
    <script …>…</script>
    <script …>…</script>
    <link rel=… …>
    <link rel=… …>
    <link rel=… …>
    <title>…</title>
    <meta …>…</meta>
    <meta …>…</meta>
    <meta …>…</meta>
    <!-- NOTE: Implied end of head, and start of body. -->
    <span itemprop="author" itemscope itemtype="http://schema.org/Person">
      <link itemprop="url" href="…">
      <link itemprop="name" content="…">
    </span>
    <span itemscope itemtype="https://schema.org/BreadcrumbList">
      <span itemprop="itemListElement" itemscope itemtype="https://schema.org/ListItem">
        <meta itemprop="position" content="1"/>
        <span itemprop="item" itemid="…" itemscope itemtype="https://schema.org/Thing">
          <meta itemprop="name" content="…"/>
        </span>
      </span>
    </span>
    <!-- NOTE: Declared end of head, and declared start of body. -->
  </head>
  <body>
    <ytd-app>…</ytd-app>
  </body>
</html>
```

The good news is that, some time in the last two years, YouTube moved their `<link rel="alternate" type="application/rss+xml"  …>` element up a few lines, so it is now before implied break/end of the `<head>` and working fine with NetNewsWire.
